libsolidity/analysis/ControlFlowAnalyzer.cpp

```diff
@@ -191,6 +247,100 @@ void ControlFlowAnalyzer::checkUnreachable(CFGNode const* _entry, CFGNode const*
 		// Extend the location, as long as the next location overlaps (unreachable is sorted).
 		for (; it != unreachable.end() && it->start <= location.end; ++it)
 			location.end = std::max(location.end, it->end);
-		m_errorReporter.warning(5740_error, location, "Unreachable code.");
+
+		if (m_previousUnreachable.emplace(location).second)
+			m_errorReporter.warning(5740_error, location, "Unreachable code.");
 	}
 }
+
+bool ControlFlowAnalyzer::checkForReverts(ContractDefinition const& _contract, FunctionCall const& _functionCall)
+{
+	auto const& functionType = dynamic_cast<FunctionType const&>(
+		*_functionCall.expression().annotation().type
+	);
+
+	if (!functionType.hasDeclaration())
+		return false;
+
+	auto const& unresolvedFunctionDefinition =
+		dynamic_cast<FunctionDefinition const&>(functionType.declaration());
+
+	FunctionDefinition const* functionDefinition = nullptr;
+	ContractDefinition const* contractDefinition = nullptr;
+
+	SimpleASTVisitor visitor([&](ASTNode const& _node) {
+		if (auto const* memberAccess = dynamic_cast<MemberAccess const*>(&_node))
+		{
+			if (*memberAccess->annotation().requiredLookup == VirtualLookup::Super)
+			{
+				if (auto const typeType = dynamic_cast<TypeType const*>(memberAccess->expression().annotation().type))
+					if (auto const contractType = dynamic_cast<ContractType const*>(typeType->actualType()))
+					{
+						solAssert(contractType->isSuper(), "");
+						ContractDefinition const* superContract = contractType->contractDefinition().superContract(_contract);
+
+						functionDefinition =
+							&unresolvedFunctionDefinition.resolveVirtual(
+								_contract,
+								superContract
+							);
+					}
+			}
+			else
+			{
+				solAssert(*memberAccess->annotation().requiredLookup == VirtualLookup::Static, "");
+				functionDefinition = &unresolvedFunctionDefinition;
+			}
+			return false;
+		}
+		else if (auto const* identifier = dynamic_cast<Identifier const*>(&_node))
+		{
+			solAssert(*identifier->annotation().requiredLookup == VirtualLookup::Virtual, "");
+			functionDefinition = &unresolvedFunctionDefinition.resolveVirtual(_contract);
+
+			return false;
+		}
+		return true;
+	}, [](ASTNode const&){});
+
+	_functionCall.expression().accept(visitor);
```

Comment:
Except for the last example (which should work just normal I think) I wouldn't even know how to begin to resolve those cases and @ekpyron told me in the chat that those cases are impossible to resolve for the general case.

I think it's fine if we have some complex calls that slip through the analysis.

Reply:
I would actually say that all complex calls should slip through the analysis, because that makes the compiler more predictable.

Reply:
Well, my point here was that using an AST visitor to determine which function the call expression refers to is a pretty complex way to do it and still fails in cases like these. If we cannot handle them correctly then something simpler might be better - i.e. just checking if the call expression is `Identifer` or `MemberAccess` and otherwise just assuming it's a pointer.

BTW, if these actually defeat the flow analyser, I think it might be good to add them as test cases with comments explaining that it's not supposed to handle them correctly.